### ISSUE TITLE
Gate isn't calling function if method name and ability aren't the same

### ISSUE THREAD

**Issue by ES-Vodoo** [laravel/framework#20608]
- Laravel Version: 5.4.33
- PHP Version: 7.0.10
- Database Driver & Version: n/a

```
class AuthServiceProvider extends ServiceProvider
{
    protected $policies = [
        'App\Models\Ticket' => 'App\Policies\TicketPolicy',
    ];

    public function boot()
    {
        $this->registerPolicies();

        Gate::define('test', 'TicketPolicy@example');
    }
}
```
```
class TicketPolicy
{
    use HandlesAuthorization;

    public function example($user, $ticket)
    {
        dd('ticket policy');
    }
}
```
Part of controller's function (`$ticket` is an instance of `App\Models\Ticket`)
```
if (Gate::allows('test', $ticket)) 
{
    dd(true);
}
```
As you can see I used `dd()` function to make sure Gate is working. In this case there is `test` ability and `example()` method, when I entered the page I saw ticket's view instead of `"ticket policy"`.

For several hours I was trying to fix this and then I tried to change ability name to the same as method name.

Controller:
`if (Gate::allows('example', $ticket))`

AuthServiceProvider:
`Gate::define('example', 'TicketPolicy@example');`

After these changes everything worked (`"ticket policy"` appeared). It's a bug or intended behavior?

**Comment by browner12** [comment]
you are probably running into the fun issue where the Policies are overly greedy and take precedence over the defined gates.  Since you are passing a `Ticket` model into the `allows` method, Laravel will automatically use your policy and ignore your gate definition.

 checkout this method:

https://github.com/laravel/framework/blob/5.4/src/Illuminate/Auth/Access/Gate.php#L346

and you'll see that since we are using the policy, it simply translates the ability name to the method name.

I have this fixed in master with this PR:

https://github.com/laravel/framework/pull/19120

which will allow the code to fall through to the gates if the method does not exist on the policy.

**Comment by ES-Vodoo** [comment]
Everything like before, still not working but I see small difference, after update Laravel throws `ReflectionException (Class TicketPolicy does not exist)`. If ability name and method name are the same there is no error.

Laravel Version: 5.5-dev

**Comment by themsaid** [comment]
@ES-Vodoo you have an error in your code, you need to use `Gate::define('test', UserPolicy::class.'@example');`

**Comment by ES-Vodoo** [comment]
@themsaid so there is mistake in docs, please change it because it's misleading.

Example from docs:
```
/**
 * Register any authentication / authorization services.
 *
 * @return void
 */
public function boot()
{
    $this->registerPolicies();

    Gate::define('update-post', 'PostPolicy@update');
}
```

**Comment by browner12** [comment]
@themsaid 's example is simply using the `::class` notation, which will provide you with the fully qualified namespace of the policy. You can use the `::class` notation, or you can write out the full namespace as a string.

I feel like even using just the classname should work, though, because the Gate tries to resolve policies using `$this->container->make($class)`.  Or does `make` need a FQN as well?

**Comment by ES-Vodoo** [comment]
@browner12 @themsaid's example worked for me but now `before` method is not being called. I tried to find out what's going on and I found the problem. 

It refers to this method: https://github.com/laravel/framework/blob/master/src/Illuminate/Auth/Access/Gate.php#L435

`callPolicyBefore` which is responsible for calling `before` (if exists on the policy) is not being called because mentioned method returns false (first if-statement). `formatAbilityToMethod` doesn't work properly with abilities registered using "Resource Gates" due to special format of it (eg. `posts.view`). 

It should check the method associated with ability instead of checking if method named the same as ability exists.

Hope I explained it clearly.
